### ISSUE TITLE
XGBoost Prediction returns same top features and predictive values for all cell subtypes

### ISSUE THREAD

**Issue by meitoru-laxusmimi** [yizhak-lab-ccg/PRECISE-#2]
Hi,
I'm running PRECISE on a scRNA-seq dataset and trying to compare the top features between different cell subtypes. To save computational energy, I ran Boruta Analyser for feature selection on all data and trained the XGBoost prediction model on each of the cell subtypes. However, I found the sample prediction, top features and AUC are **exactly** the same for each cell type. I have tried on different datasets, including a TNBC dataset and a pancancer dataset, and it showed the same manner. The number of cells is different for each cell subtype, so I'm sure it's not using the same dataset. Below is a sample of the results:
For CD8T:

<img width="1216" height="354" alt="Image" src="https://github.com/user-attachments/assets/390e3685-2244-4656-b3f0-9c753dfe4b3f" />

For B Plasma Cells: 
<img width="578" height="176" alt="Image" src="https://github.com/user-attachments/assets/032991b3-b5c5-449c-b4ab-ca4c184136b5" />

And below is the code I ran for comparing each cell type: 

cell_types_list = ["CD8T", "CD4T", "NKT", "MAIT", "B_Plasma_cells", "Mast", "Megakaryocyte", "Macrophage", "Monocyte", "DC"]
results_all_celltype_tumour = []
for cell_type in cell_types_list:
    print(cell_type)
    cell_type_tumour_data = set(adata_toprocess_tumour_filtered.obs["SubTypes"])
    if cell_type not in cell_type_tumour_data:
        continue
    else:
        output_dir = "../tumour_results_celltype_comparison/" + cell_type + "/"
        precise = Precise(
            adata_toprocess_tumour_filtered,
            model_name='XGBoostClassifier',
            output_dir=output_dir,
            target_column="response",  # Adjust based on your dataset
            sample_column="patient_time",
            celltype=cell_type, # is that right?
            args=args
        )
        prediction_results = precise.cv_prediction()
        sample_results, auc, estimators = prediction_results
        results = {"cell_type": cell_type, "sample_results": sample_results, "auc": auc}
        results_all_celltype_tumour.append(results)
        precise.create_feature_plots(1000)
        output_dir_shap = os.path.join(output_dir,cell_type)
        if not os.path.exists(output_dir_shap):
            os.makedirs(output_dir_shap)
        precise.run_shap_visualizations(save_path=output_dir_shap)
        precise.run_shap_visualizations()
        
 Your training and testing results are looking promising, so I'm quite curious why this is happening. Do you also have any suggestions on hyperparameter tuning on this model as well? Thank you.
Yuheng

**Comment by asafpinhasi** [comment]
Hi Yuheng,

The main issue seems to be with how the cell type is provided as input. PRECISE expects each cell type (e.g., “T cell”) to be supplied as a separate boolean or 0/1 column in .obs, rather than having a single column (e.g., "SubType") with cell types as categorical values. This design allows specifying overlapping cell types when needed. Going over the code I've found a different bug that could cause issues there, so please reclone the repo. 
On the example data, it will look as follows:
```
import sys
import os
sys.path.append(os.path.abspath('../../PRECISE-/modules/'))
from Precise import Precise
import anndata as ad

adata = ad.read_h5ad("../examples/sample_melanoma_adata.h5ad")
precise = Precise(
        adata,
        model_name="XGBoostClassifier",
        output_dir=f"./results",
        target_column="response",
        sample_column="sample_name",
    )

results = {}
for celltype in ["B cells", "T cells"]:
    print(f"Running PRECISE for cell type: {celltype}")
    precise = Precise(
        adata,
        model_name="XGBoostClassifier",
        output_dir=f"./results",
        target_column="response",
        sample_column="sample_name",
        celltype=celltype,
    )
    sample_results, auc, estimators = precise.cv_prediction(verbose=True)
    results[celltype] = {"auc": auc, "features": estimators}

```
I’ve now clarified this in the README, implemented more informative warnings when the cell type format is incorrect, and will soon add a section at the end of the tutorial.

Regarding hyperparameter tuning:
Some parameters (e.g., max_depth for tree-based models) can be adjusted directly by passing an args dictionary, but tuning options are currently limited. In our experience, RandomForest, XGBoost, and LightGBM generally work well out-of-the-box and do not show major gains from hyperparameter tuning on the datasets we tested.

If you want full hyperparameter flexibility, you can directly use the model wrappers and pass them to PredictionAnalyzer. For example:

```
from lightgbm import LGBMClassifier
from prediction_analysis import PredictionAnalyzer
from ML_models import SklearnModelWrapper

model = SklearnModelWrapper(
    LGBMClassifier(learning_rate=0.1, n_estimators=100, verbosity=1)
    )

prediction_analyzer = PredictionAnalyzer(
    adata,
    model=model,
    model_name='custom_LGBM_Classifier',
    target_column='response',
    sample_column='sample',
    results_folder='../results/custom_LGBM_Classifier_results',
)
prediction_analyzer.cv_prediction(k_folds=5, verbose=True)

```
Thanks for raising this!

Asaf

**Comment by meitoru-laxusmimi** [comment]
Hi Asaf, 
Thank you for the reply. I re-clone the repo, and the cell type accuracies make sense now, thank you.
I have one more question regarding the hyperparameter tuning process, since I'm quite new to machine learning. Your model used LOOCV, so the input should be a cohort - that means there are no training-validation-testing splits, as far as I understand. So, considering saving computational time, do you recommend using a portion of the samples or cells for each sample to do hyperparameters tuning? 
Thanks,
Yuheng

**Comment by asafpinhasitechnion** [comment]
Hi Yuheng,

You’re absolutely right, a major limitation of LOOCV is that it doesn’t really allow hyperparameter tuning (or at least it makes it extremely computationally intensive).

To be rigorous, the splitting must always be done per sample, not per cell, since using cells from the same sample for both training and validation would leak information.

Because the number of samples is small, robust hyperparameter tuning is challenging. If possible, the best option is to use a larger cohort, like the one we used in the discussion section of our manuscript (link to data - https://cellxgene.cziscience.com/collections/61e422dd-c9cd-460e-9b91-72d9517348ef). With more samples, you could split them into train/validation/test (for example, in a k-fold CV setting), which makes tuning much more reliable.

If you need to use the smaller cohort, one approach is to use nested cross-validation. For example: Split the data 5 times into train/test folds. Within each train fold, run another 5-fold CV across different hyperparameter sets to pick the best ones. Use these parameters to train on the whole training set in the split, and predict on the held-out samples. This isn’t ideal, but it avoids data leakage and should give reasonable results given the sample size.

By the way, you can use PredictionAnalyzer.cv_prediction to run k-fold cv not only LOOCV, by passing k_folds as a parameter.

Thanks,

Asaf

**Comment by meitoru-laxusmimi** [comment]
Hi Asaf,
I really appreciate your response! I'm running PRECISE with two datasets that are 55 samples and 195 samples. So I guess I'll try with both. 
Furthermore, according to your README file, after running reinforcement learning using the top features selected by Boruta, the next step is to run the shap and run prediction. Is this for running with the original anndata or the refined anndata? (I assumed it's the refined anndata?)
Cheers,
Yuheng

**Comment by asafpinhasitechnion** [comment]
Hi Yuheng,

The larger dataset already seems reasonable to have one train-val-test split per fold in a k-fold cross-validation, especially if you are concerned about run time.
Regarding the SHAP analysis, it really depends on your needs. Running SHAP on the refined anndata will be much faster and more focused. If you want a broader understanding of how the different features relate to each other and to response in a wider context, using the original anndata is more appropriate. In our study we used the full anndata.
Just a note – the RL step, SHAP, and prediction are independent. They’re included to demonstrate functionality rather than being strictly sequential. Also, since RL uses the labels, you should be careful not to let information from this step leak into later prediction tasks.

Best,
Asaf

**Comment by meitoru-laxusmimi** [comment]
Hi Asaf,
Thank you for all the help. I have one more question regarding the validation dataset. Did you involve all cell types, including malignant cells, for your validation analysis? (For this type of study, should we involve malignant cells?)
Yuheng

**Comment by asafpinhasitechnion** [comment]
Hi Yuheng 🙂

Apologies for the late reply.
We used T cells in the validation dataset, as many cohorts focused only on subsets of the data, and T cells were the most consistently available cell type.
We did not include malignant cells at any stage of the study. In general, learning common response-related features from malignant cells is challenging due to the high heterogeneity of tumors. Tumors can show very different expression patterns across patients, and even between samples from the same patient (for example, see Figure 2 in https://www.nature.com/articles/s41588-024-01997-z , which clearly illustrates patient-specific malignant-cell clustering).
Immune cells, on the other hand, tend to share more common patterns between samples and even across cancer types, making it easier for the model to learn generalizable features. That’s not to say this can’t be done with malignant cells, but doing so would likely require a much larger number of samples to capture the underlying variability.

If possible, I’d recommend focusing on T cells in this validation dataset, and in any case, I’d avoid including malignant cells and would limit the analysis to immune cells.

Best,
Asaf
